Added new generate mode for generating resources only with xliffs

gouniLee opened the pull request:
I've added a new `generate` mode to generate resource only with cliffs without scanning sources.
The webOS needs a feature that can generate resources without scanning source files, only refers xliff files.
it creates JSON resource styles named 'strings.json as a default. but it provides the option to set a resource directory and directory name.
Hope to get a review of the approaches and value names etc.
usage:
`> node loctool.js generate -x [xliffsPath]  --xliffResRoot [path] --xliffResName [name]`

Added new generate mode for generating resources only with xliffs

ehoogerbeets commented on lib/GenerateResource.js:
```diff
@@ -0,0 +1,191 @@
+    this.ts = new TranslationSet(this.sourceLocale);
+};
+
+var xliffFileFilter = /(^|[^a-z])([a-z][a-z][a-z]?)(-([A-Z][a-z][a-z][a-z]))?(-([A-Z][A-Z]))?\.xliff$/;
```
There may be a prefix to the xliff file name, like "myapp-ko-KR.xliff", so this should be updated

ehoogerbeets commented on lib/GenerateResource.js:
```diff
@@ -0,0 +1,191 @@
+        }.bind(this));
+    }
+
+    if (this.pathName) {
```
`this.pathName` is not set anywhere in this class. Maybe it should be in the constructor?

gouniLee replied:
It's accidentally copied from another file. I removed it because it's valid only xliff 1.0.which doesn't support generate mode.

ehoogerbeets commented on lib/GenerateResource.js:
```diff
@@ -0,0 +1,191 @@
+    }
+    var fullPath;
+    for (locale in this.output){
+        var resourcePath = this._calResourcePath(locale);
```
This is hard-coded to write out json files in webos format. Each resource identifies what type it is and what class handles it. This code should get the resource file type for each resource, add the resource to it, and then ask each resource file type to write itself out. That way, `generate` can be used with any project, not just webos projects.

gouniLee replied:
Yeah, I removed codes here. then made it to generate the file through plugins. plugin info will be extracted project configuration. the config file is an essential file for localization. so I think it's ok to use that info.
and I've defined newly process to generate localization which omitted generatePseudo, extract and write task.
https://github.com/iLib-js/loctool/blob/createGenerateMode/loctool.js#L466

ehoogerbeets replied:
I don't see those changes here... are they pushed to github yet?

ehoogerbeets commented on loctool.js:
```diff
@@ -83,6 +83,10 @@ function usage() {
         "--localizeOnly\n" +
         "  Generate a localization resource only. Do not create any other files at all after running loctool. \n" +
+        "--xliffResRoot\n" +
+        "  Specify the dir where the xliffs resource generation output (Default is resources/) live.. \n" +
```
Make this `Specify the dir where the generation output should go. (Default is resources/)`

gouniLee replied:
Update it. Thanks!

ehoogerbeets commented on loctool.js:
```diff
@@ -83,6 +83,10 @@ function usage() {
+        "--xliffResRoot\n" +
+        "  Specify the dir where the xliffs resource generation output (Default is resources/) live.. \n" +
+        "--xliffResName\n" +
+        "  Specify the resource filename from xliff resource generation (Default is strings.json/) live.. \n" +
```
Make this `Specify the resource filename used during resource file generation. (Default is strings.json/)`

gouniLee replied:
Update it. Thanks!

ehoogerbeets commented on loctool.js:
```diff
@@ -453,6 +462,16 @@ try {
+            resRoot: settings.xliffResRoot,
+            resName: settings.xliffResName
+        });
+        genResource.init();
```
`init()` and `generate()` should be called separately so that they can be unit tested independently

ehoogerbeets commented on test/testGenerateResource.js:
```diff
@@ -0,0 +1,84 @@
+        test.equal(genres.getResSize(), 2);
+
+        test.done();
+    },
```
Need tests here to test that `init()` read the resource files properly, and that `generate()` wrote the resource files properly.

gouniLee replied:
I've separated to two functions and added test cases too.

ehoogerbeets commented on package.json:
```diff
@@ -1,6 +1,6 @@
     "name": "loctool",
-    "version": "2.8.2",
+    "version": "2.9.0",
     "main": "./loctool.js",
```
This is a new feature, so the version needs to be changed to 2.10.0

gouniLee commented:
I've updated a bit more. It allows for generating resources when the `project.json` doesn't exist. In order to do that, I've newly defined additional parameters (`--projectType`, `plugins`) and fixed code to work `Customproject` properly that I expected. Because the process of `extract` and `write` are skipped in generate mode,  the code that creates the files in `generateResource` class is included. the default is strings.json

gouniLee commented:
closing this PR. and recreated https://github.com/iLib-js/loctool/pull/119 in order to see the changes clearly.
